### ISSUE TITLE
Proposal: Creating a Python class out of a Julia type

### ISSUE THREAD

**Issue by cstjean** [stevengj/PyCall.jl#235]
I would like to pass my Julia objects as Python objects with methods defined in Julia (similar to #138). I was happy to see that it's possible through `pyjlwrap_type`, even if it's not a very straight-forward endeavor (unless I've missed some functionality somewhere).

I think I can implement a more high-level interface, but I'd like to run the proposal by you before making a PR. The interface would look like

```julia
type IO
   ...
end

function jl_IO_readline(io::IO, nline)
   d = readline(io)
   if length(d) > nb
       resize!(d, nb)
   end
   return d
end

...

@def_py_methods IO, ("readline", jl_IO_readline), ...
```

where the macro will expand into 
- a `pyjlwrap_type` call to create the new type with `PyMethodDef(..., METH_VARARGS)`
- a `PyObject(io::IO) = pyjlwrap_new(wrapped_type, io)` method definition
- a `jl_IO_readline(self_::PyPtr, args_::PyPtr)` definition that uses `convert(PyAny)` to read all its arguments, then forwards them to the correct method, similar to the callback.jl definition `PyObject(f(convert(PyAny, args)...))`.

Does that sound reasonable?

**Comment by stevengj** [comment]
Rather than defining a new `jl_IO_readline` method, it would probably be better to use `gensym` to define an "anonymous" method.  (Actually, macro hygiene may do this for you.)

Would be good to also define a way to define attribute getters/setters (e.g. this is required for the "closed" attribute of the Python IO type).

We also need a way to declare that the type is a subtype of some other Python type.     

Maybe a more Julian syntax:
```jl
@pydef type IO <: somepythontype
     readline(io, nline) = jl_IO_readline(io, nline)
     multiply(io, x, y) = x * y
     closed.get(io) = !isopen(io)
     myattribute.set!(io, val) = (io.myattribute = val)
end
```

**Comment by cstjean** [comment]
Thank you for your comments. That's a bit more ambitious than I intended, but I'll see how many of these I can implement. Is there an example of inheritance in PyCall?

I'm beginning to wrap my head around how PyCall works. I have a question about `PyCall.jl`:

```julia
function pydecref(o::PyObject)
    ccall(@pysym(:Py_DecRef), Void, (PyPtr,), o.o)
    o.o = convert(PyPtr, C_NULL)
    o
end

function pyincref(o::PyObject)
    ccall((@pysym :Py_IncRef), Void, (PyPtr,), o)
    o
end
```
`pydecref` makes sense, but why is `Py_IncRef` called on `o`, and not `o.o`?

**Comment by stevengj** [comment]
There aren't any examples of Python inheritance in the PyCall source, but to do it all you need to do is to set the [`tp_base` field](https://docs.python.org/2/c-api/typeobj.html#c.PyTypeObject.tp_base) of `PyTypeObject` to point to the desired Python type.

Regarding your second questions, there is an [`unsafe_convert` method](https://github.com/stevengj/PyCall.jl/blob/master/src/PyCall.jl#L111) so that if you pass a `PyObject` to a `PyPtr` argument it automatically passes the `.o` field.

**Comment by cstjean** [comment]
Solved by #250
